src/lib/rng/hmac_rng/hmac_rng.cpp

```diff
@@ -12,41 +12,60 @@
 
 namespace Botan {
 
-HMAC_RNG::HMAC_RNG(const std::string& hash, size_t max_output_before_reseed) :
-   Stateful_RNG(max_output_before_reseed)
+HMAC_RNG::HMAC_RNG(std::unique_ptr<MessageAuthenticationCode> prf,
+                   RandomNumberGenerator& underlying_rng,
+                   Entropy_Sources& entropy_sources,
+                   size_t max_output_before_reseed) :
+   Stateful_RNG(underlying_rng, max_output_before_reseed),
+   m_prf(std::move(prf)),
+   m_extractor(m_prf->clone())
    {
-   m_extractor = MAC::create("HMAC(" + hash + ")");
-   if(!m_extractor)
-      throw Invalid_Argument("HMAC_RNG hash not found");
+   if(!m_prf->valid_keylength(m_prf->output_length()))
+      {
+      throw Invalid_Argument("HMAC_RNG cannot use " + m_prf->name());
+      }
 
-   m_prf.reset(m_extractor->clone());
+   this->clear();
+   }
 
-   if(!m_prf->valid_keylength(m_extractor->output_length()) ||
-      !m_extractor->valid_keylength(m_prf->output_length()))
+HMAC_RNG::HMAC_RNG(std::unique_ptr<MessageAuthenticationCode> prf,
+                   RandomNumberGenerator& underlying_rng,
+                   size_t max_output_before_reseed) :
+   Stateful_RNG(underlying_rng, max_output_before_reseed),
+   m_prf(std::move(prf)),
+   m_extractor(m_prf->clone())
+   {
+   if(!m_prf->valid_keylength(m_prf->output_length()))
       {
-      throw Invalid_Argument("HMAC_RNG: Bad algo combination " +
-                             m_extractor->name() + " and " +
-                             m_prf->name());
+      throw Invalid_Argument("HMAC_RNG cannot use " + m_prf->name());
       }
 
    this->clear();
    }
 
-/*
-* HMAC_RNG Constructor
-*/
-HMAC_RNG::HMAC_RNG(MessageAuthenticationCode* extractor,
-                   MessageAuthenticationCode* prf,
+HMAC_RNG::HMAC_RNG(std::unique_ptr<MessageAuthenticationCode> prf,
+                   Entropy_Sources& entropy_sources,
                    size_t max_output_before_reseed) :
-   Stateful_RNG(max_output_before_reseed),
-   m_extractor(extractor), m_prf(prf)
+   Stateful_RNG(entropy_sources, max_output_before_reseed),
+   m_prf(std::move(prf)),
+   m_extractor(m_prf->clone())
+   {
+   if(!m_prf->valid_keylength(m_prf->output_length()))
+      {
+      throw Invalid_Argument("HMAC_RNG cannot use " + m_prf->name());
+      }
+
+   this->clear();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this call from a constructor/destructor to the overridable "clear" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS1699)